File: ruby/robot-name/robot_name.rb
class Robot

  attr_reader :name

  def initialize
    @name = name_generator
  end

  def reset
    @name = name_generator
  end

  private

  def name_generator
    letters = ('A'..'Z').to_a.sample(2)
    numbers = (0..9).to_a.sample(3)
    (letters + numbers).join
  end
end

File: javascript/robot-name/robot-name.js
Robot = function() {

  this.name = this.generateName()

  this.reset = function() {
    this.name = this.generateName()
  }
}

var letters = "ABCDEFGHIJKLMNOPQRSTUVWXYZ".split('')
var numbers = "0123456789".split('')

var rand = function(array) {
  var index = Math.floor(Math.random() * array.length)
  return array[index];
}

var names = []

Robot.prototype.generateName = function() {
  var name = ""
  name += rand(letters);
  name += rand(letters);
  name += rand(numbers);
  name += rand(numbers);
  name += rand(numbers);

  if (names.indexOf(name) === -1) {
    names.push(name)
    return name;
  } else {
    return this.generateName()
  }
}

module.exports = Robot;
